metrics/prometheus.go

```diff
@@ -0,0 +1,261 @@
+			), // "connection_error" = {"accept", "close"}
+		},
+		ExtraTTL: &PrometheusExtraTTLMetrics{
+			ExtraTTLSeconds: newHistogram(cfg, registry,
+				"extra_ttl_seconds",
+				"Extra time to live in seconds specified",
+				cacheWriteTimeBuckts,
+			),
+		},
+	}
+	promMetrics.ExtraTTL.ExtraTTLSeconds.Observe(5000.00)
+
+	return promMetrics
+}
+
+/**************************************************
+ *	Helper Init functions
+ **************************************************/
+func newCounterVecWithLabels(cfg config.PrometheusMetrics, registry *prometheus.Registry, name string, help string, labels []string) *prometheus.CounterVec {
+	opts := prometheus.CounterOpts{
+		Namespace: cfg.Namespace,
+		Subsystem: cfg.Subsystem,
+		Name:      name,
+		Help:      help,
+	}
+	counterVec := prometheus.NewCounterVec(opts, labels)
+	registry.MustRegister(counterVec)
+	return counterVec
+}
+
+func newSingleCounter(cfg config.PrometheusMetrics, registry *prometheus.Registry, name string, help string) prometheus.Counter {
+	opts := prometheus.CounterOpts{
+		Namespace: cfg.Namespace,
+		Subsystem: cfg.Subsystem,
+		Name:      name,
+		Help:      help,
+	}
+	counter := prometheus.NewCounter(opts)
+	registry.MustRegister(counter)
+	return counter
+}
+
+func newHistogram(cfg config.PrometheusMetrics, registry *prometheus.Registry, name, help string, buckets []float64) prometheus.Histogram {
+	opts := prometheus.HistogramOpts{
+		Namespace: cfg.Namespace,
+		Subsystem: cfg.Subsystem,
+		Name:      name,
+		Help:      help,
+		Buckets:   buckets,
+	}
+	histogram := prometheus.NewHistogram(opts)
+	registry.MustRegister(histogram)
+	return histogram
+}
+
+func newGaugeMetric(cfg config.PrometheusMetrics, registry *prometheus.Registry, name string, help string) prometheus.Gauge {
+	opts := prometheus.GaugeOpts{
+		Namespace: cfg.Namespace,
+		Subsystem: cfg.Subsystem,
+		Name:      name,
+		Help:      help,
+	}
+	gauge := prometheus.NewGauge(opts)
+	registry.MustRegister(gauge)
+	return gauge
+}
+
+func newHistogramVector(cfg config.PrometheusMetrics, registry *prometheus.Registry, name, help string, labels []string, buckets []float64) *prometheus.HistogramVec {
+	opts := prometheus.HistogramOpts{
+		Namespace: cfg.Namespace,
+		Subsystem: cfg.Subsystem,
+		Name:      name,
+		Help:      help,
+		Buckets:   buckets,
+	}
+	histogram := prometheus.NewHistogramVec(opts, labels)
+	registry.MustRegister(histogram)
+	return histogram
+}
+
+/**************************************************
+ *	DEPECRATED Functions to record metrics
+ **************************************************/
+func (m PrometheusMetrics) Export(cfg config.Metrics) {
+}
+
+/**************************************************
+ *	NEW Functions to record metrics
+ **************************************************/
+func (m *PrometheusMetrics) RecordPutRequest(status string, duration *time.Time) {
+	incCounterInVector(m.Puts.RequestStatus, "status", status, []string{"add", "error", "bad_request"})
```

Comment:
I don't understand what the slice at the end is doing.

Reply:
Because `PrometheusMetrics`, `PrometheusRequestStatusMetric`, `PrometheusRequestStatusMetricByFormat`, and `PrometheusConnectionMetric` are all objects with `prometheus.CounterVec` fields, a single function to increment their fields was created. Therefore, if the `m.Connections.ConnectionsErrors` `counterVec` has the labels: "accept_error" and "close_error", we call `incCounterInVector` like:
```
incCounterInVector(m.Connections.ConnectionsErrors, "connection_error", label, []string{"accept_error", "close_error"})
```
`m.PutsBackend.PutBackendRequests` on the other hand, has more than two labels, so this counter gets incremented like:
```
incCounterInVector(m.PutsBackend.PutBackendRequests, "label", label, []string{"add", "json", "xml", "invalid_format", "defines_ttl", "error"})
```
As a matter of fact a map[string]bool makes a lot more sense. I'll modify this.